lib/schema.js

```diff
@@ -222,49 +376,105 @@ class Metaschema {
+        errors.push(
+          new ValidationError('unresolvedProperty', `${path}${prop}`)
+        );
         continue;
       }
-      const def = definition[key];
-      const val = instance[key];
-      const type = typeof val;
-      const domain = this.domains.get(def.domain);
-      if (!insKey && defKey) {
-        if (typeof def === 'function') {
-          const decorator = extractDecorator(def.prototype);
-          if (decorator === 'Validate' && !def(instance)) {
-            errors.push('Validation failed');
-            continue;
-          }
+
+      const definition = schema[prop];
+
+      if (extractDecorator(definition) === 'Validate' && !patch) {
+        if (!definition(object)) {
+          errors.push(new ValidationError('validation', `${path}${prop}`));
+        }
+        continue;
+      }
+
+      if (definition.readOnly && patch) {
+        errors.push(new ValidationError('immutable', `${path}${prop}`));
+        continue;
+      }
+
+      const value = object[prop];
+
+      if (!isObjectProp) {
+        if (definition.required && !patch) {
+          errors.push(new ValidationError('missingProperty', `${path}${prop}`));
         }
-        if (def.required) {
-          errors.push(`Field ${key} not found`);
+        continue;
+      }
+
+      if (value === undefined || value === null) {
+        if (definition.required) {
+          errors.push(new ValidationError('emptyValue', `${path}${prop}`));
         }
         continue;
       }
-      if (domain && type !== domain.type) {
-        errors.push(`Field ${key}:${val} expected to be ${domain.type}`);
+
+      if (definition.domain) {
+        errors.push(
+          ...validateDomain(value, `${path}${prop}`, definition.definition)
+        );
+      } else if (definition.category) {
+        errors.push(
+          ...validateLink(value, `${path}${prop}`, definition, this, patch)
+        );
       }
     }
+    return errors;
+  }
+
+  // Validate instance against a category
+  //   categoryName - <string> category name
+  //   instance - <any> instance to validate
+  //   patch - <boolean>, flag to determine if the object contains patch or
+  //       value, default: `false`
+  //   path - <string>, path to an object, for nested objects, default: `''`
+  // Returns: <MetaschemaError> | <null>
+  validateCategory(categoryName, instance, patch = false, path = '') {
+    const category = this.categories.get(categoryName);
+    if (!category) {
+      return new MetaschemaError(
+        [new ValidationError('undefinedEntity', null, 'category')]
+      );
+    }
+
+    const errors = this[validate](category.definition, instance, patch, path);
+
+    return errors.length === 0 ? null : new MetaschemaError(errors);
+  }
+
+  // Validate instance against a category
+  //   formName - <string> form name
+  //   args - <any> form arguments
+  // Returns: <MetaschemaError> | <null>
+  validateForm(formName, args) {
+    const form = this.categories.get(formName);
```

Comment:
Should this maybe access `this.forms` instead of `this.categories`?
```suggestion
    const form = this.forms.get(formName);
```